app/src/main/java/uw/hcrlab/kubi/robot/Robot.java

```diff
@@ -187,23 +321,111 @@ public void startup() {
     public void shutdown() {
         Log.i(TAG, "Shutting down Main Thread ...");
 
-        while (true) {
-            try {
-                if(App.InWizardMode()) {
-                    questions.Stop();
-                }
+        if (sleep != null) {
+            sleep.cancel();
+        }
 
-                if(thread != null) {
-                    thread.setRunning(false);
-                    thread.join();
-                    thread = null;
-                }
+        if (bored != null) {
+            bored.cancel();
+        }
+
+        unloadAllPronunciations();
+
+        progress.cleanup();
+
+        if (App.InWizardMode()) {
+            questions.Stop();
+        }
+
+//        while (true) {
+//            try {
+//                if(App.InWizardMode()) {
+//                    questions.Stop();
+//                }
+//
+//                if(thread != null) {
+//                    thread.setRunning(false);
+//                    thread.join();
+//                    thread = null;
+//                }
+//
+//                return;
+//            } catch (InterruptedException e) {
+//                Log.e(TAG, "Robot thread didn't join. Trying again.");
+//            }
+//        }
+    }
+
+    private String normalizeText(String text) {
+        // Normalize the text
+        text = text.toLowerCase();
+        text = text.replace(',',' ');
+        text = text.replace('.',' ');
+        text = text.trim();
+        return text;
+    }
+
+    public void loadPronunciation(String text) {
+        text = normalizeText(text);
+
+        String url = App.getAudioURL(text);
+
+        if (url != null) {
+            String audioUrl = proxy.getProxyUrl(url);
+            mPronunciations.put(text, MediaPlayer.create(mActivity, Uri.parse(audioUrl)));
+        }
+    }
+
+    public void unloadPronunciation(String text) {
+        text = normalizeText(text);
+
+        MediaPlayer mp = mPronunciations.remove(text);
 
-                return;
-            } catch (InterruptedException e) {
-                Log.e(TAG, "Robot thread didn't join. Trying again.");
+        if (mp != null) {
+            mp.release();
+        }
+    }
+
+    public void unloadAllPronunciations() {
+        for (Map.Entry kvp : mPronunciations.entrySet()) {
+            MediaPlayer mp = (MediaPlayer) kvp.getValue();
+            mp.release();
+        }
+
+        mPronunciations.clear();
+    }
+
+    public void pronounceAfterSpeech(String text) {
+        delayedPronunciation = text;
+    }
```

Comment:
Yes. Absolutely. I think we actually want a queue of TTS utterances and pronunciations which can be mixed up in any order. I just didn't get that far. I was thinking it would be something like (pseudo-code):

    robot.sayInOrder({
      {some TTS},
      {some Pronunciation},
      {some TTS}
    })